### PR TITLE
Raftstore-v2: transfer leader

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #12842

What's Changed:

```
Transfer leader is similar to v1.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore-v2/src/fsm/peer.rs [hunk 1]
```diff
@@ -127,6 +129,18 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport> PeerFsmDelegate<'a, EK, ER,
         Self { fsm, store_ctx }
     }
 
+    #[inline]
+    fn schedule_pending_ticks(&mut self) {
+        let pending_ticks = self.fsm.peer.take_pending_ticks();
+        for tick in pending_ticks {
+            if tick == PeerTick::ReactivateMemoryLock {
+                self.register_reactivate_memory_lock_tick()
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/fsm/peer.rs:137]: what's the difference of register_reactivate_memory_lock_tick and self.schedule_tick(ReactivateMemoryLock)? if it's same, why bother to differentiate it here?

File: components/raftstore-v2/src/fsm/peer.rs [hunk 2]
```diff
@@ -245,3 +262,37 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport> PeerFsmDelegate<'a, EK, ER,
...
+        {
+            pessimistic_locks.status = LocksStatus::Normal;
+            self.fsm.reactivate_memory_lock_ticks = 0;
+        } else {
+            drop(pessimistic_locks);
+            self.register_reactivate_memory_lock_tick();
+        }
+    }
+
+    pub fn register_reactivate_memory_lock_tick(&mut self) {
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/fsm/peer.rs:295]: no need to have a method if it's just schedule_tick.

File: components/raftstore-v2/src/operation/command/admin/leader_transfer.rs [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+                        // Set this flag to propose this command like a normal proposal.
+                        cmd.mut_header()
+                            .set_flags(WriteBatchFlags::TRANSFER_LEADER_PROPOSAL.bits());
+                        cmd.mut_admin_request()
+                            .set_cmd_type(AdminCmdType::TransferLeader);
+                        cmd.mut_admin_request().mut_transfer_leader().set_peer(from);
+                        if let (PeerMsg::RaftCommand(req), sub) = PeerMsg::raft_command(cmd) {
+                            self.on_admin_command(ctx, req.request, req.ch);
+                        } else {
+                            unreachable!();
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/command/admin/leader_transfer.rs:203]: i don't think we need unreachable!() here when on_admin_command failed. Just log error instead.

File: components/raftstore-v2/tests/integrations/test_transfer_leader.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    metapb,
+    raft_cmdpb::{AdminCmdType, CmdType, Request, TransferLeaderRequest},
+};
+use raft::prelude::ConfChangeType;
+use raftstore_v2::router::PeerMsg;
+use tikv_util::store::new_peer;
+
+use crate::cluster::Cluster;
+
+fn put_data(
```
Comment (tonyxuqqi) [inline components/raftstore-v2/tests/integrations/test_transfer_leader.rs:17]: do you have test case to verify the lock transfer logic?

File: components/raftstore-v2/src/fsm/peer.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/fsm/peer.rs:137]: Good suggestion.

File: components/raftstore-v2/src/operation/command/admin/leader_transfer.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/command/admin/leader_transfer.rs:203]: unreachable here means PeerMsg::raft_command should only return PeerMsg::RaftCommand type.

File: components/raftstore-v2/tests/integrations/test_transfer_leader.rs (hunk 4 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/tests/integrations/test_transfer_leader.rs:17]: V1's integration test has tests of it. Once we have v2's integration test, we can reuse them.

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore-v2/src/raft/peer.rs [hunk 5]
```diff
@@ -84,6 +84,7 @@ pub struct Peer<EK: KvEngine, ER: RaftEngine> {
     /// Transaction extensions related to this peer.
     txn_ext: Arc<TxnExt>,
     txn_extra_op: Arc<AtomicCell<TxnExtraOp>>,
+    pending_ticks: Vec<PeerTick>,
```
Comment (sticnarf) [inline components/raftstore-v2/src/raft/peer.rs:87]: Is it more appropriate not to put it under the "transaction extentions" section? Most ticks don't seem to be related to transactions.

File: components/raftstore-v2/src/fsm/peer.rs [hunk 6]
```diff
@@ -249,3 +262,33 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport> PeerFsmDelegate<'a, EK, ER,
         self.fsm.peer.propose_pending_writes(self.store_ctx);
     }
 }
+
+impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport> PeerFsmDelegate<'a, EK, ER, T> {
```
Comment (sticnarf) [inline components/raftstore-v2/src/fsm/peer.rs:266]: Why put it in a separate `impl` block?

File: components/raftstore-v2/src/operation/command/admin/leader_transfer.rs [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+    let req = msg.get_admin_request();
+    if !req.has_transfer_leader() {
+        return None;
+    }
+
+    Some(req.get_transfer_leader())
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    /// Return true to if the transfer leader request is accepted.
```
Comment (sticnarf) [inline components/raftstore-v2/src/operation/command/admin/leader_transfer.rs:50]: The "to" reads weird and should be removed. (I see it's copied from the original comment.)

File: components/raftstore-v2/src/operation/command/admin/leader_transfer.rs [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+                "peer" => ?peer,
+            );
+            return false;
+        }
+
+        // Broadcast heartbeat to make sure followers commit the entries immediately.
+        // It's only necessary to ping the target peer, but ping all for simplicity.
+        self.raft_group_mut().ping();
+
+        let mut msg = eraftpb::Message::new();
```
Comment (sticnarf) [inline components/raftstore-v2/src/operation/command/admin/leader_transfer.rs:138]: Here, the index is missing compared to v1. I think it is worth a TODO comment because the index is important for reducing jitter.

File: components/raftstore-v2/src/operation/command/admin/leader_transfer.rs [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+        // have marked some locks as deleted will fail because raft rejects more
+        // proposals.
+        // It is OK to return true here if it's in other states like MergingRegion or
+        // NotLeader. In those cases, the locks will fail to propose and nothing will
+        // happen.
+        if !pessimistic_locks.is_writable() {
+            return true;
+        }
+        pessimistic_locks.status = LocksStatus::TransferringLeader;
+        self.add_pending_tick(PeerTick::ReactivateMemoryLock);
```
Comment (sticnarf) [inline components/raftstore-v2/src/operation/command/admin/leader_transfer.rs:338]: `reactivate_memory_lock_ticks` is not reset.
And, what's the purpose of adding a pending tick instead of scheduling tick directly?

File: components/raftstore-v2/src/fsm/peer.rs (hunk 6 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/fsm/peer.rs:266]: I think the impl above relates to the general logic. Is it better to put them together?

File: components/raftstore-v2/src/operation/command/admin/leader_transfer.rs (hunk 7 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/command/admin/leader_transfer.rs:50]: Ok, I see.

File: components/raftstore-v2/src/operation/command/admin/leader_transfer.rs (hunk 8 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/command/admin/leader_transfer.rs:138]: Yes, that's the warm up optimization which should be implemented later. I will add todo.

File: components/raftstore-v2/src/operation/command/admin/leader_transfer.rs (hunk 9 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/command/admin/leader_transfer.rs:338]: In v1, this method is implemented under struct PeerFsmDelegate, but it is under Peer in v2. It cannot access reactivate_memory_lock_ticks and others directly.

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/command/admin/leader_transfer.rs:338]: > reactivate_memory_lock_ticks is not reset.

good catch.

File: components/raftstore-v2/src/fsm/peer.rs (hunk 6 above)
Comment (sticnarf) [reply components/raftstore-v2/src/fsm/peer.rs:266]: I don't have any preference. I was just a bit curious about it.

You can add a bit of comments to the impl block.

File: components/raftstore-v2/src/operation/command/admin/leader_transfer.rs (hunk 9 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/command/admin/leader_transfer.rs]: Yes, I added it in the `schedule_pending_ticks`

**Review by sticnarf — APPROVED** [review]
LGTM
